File: packages/agent-core/src/agent/context/tool-result-render.ts
/**
 * The single place where a stored tool result (pure data + structured
 * status/note) is rendered into the content the model actually receives.
 *
 * History and wire records store facts: the tool's own `output`, the
 * structured `isError` flag, and an optional `note` (content routed to the
 * model but never to user-facing UIs — see `ExecutableToolResult.note`).
 * Rendering those facts into model-visible text is a provider-boundary
 * concern and happens exactly once, here:
 *
 * - a failed call gets a plain-text `ERROR:` status prefix (meant for both
 *   the model and humans — UIs that show projected text may show it too);
 * - an empty output is replaced with a placeholder so strict providers do
 *   not reject an empty tool message;
 * - the note, when present, is appended verbatim. No wrapping is added: any
 *   formatting is the producing tool's choice. A text-only result keeps a
 *   SINGLE text part (note joined with a newline): providers serialize that
 *   as plain string tool content — some OpenAI-compatible backends reject
 *   content-part arrays on tool messages, and joining providers (Google
 *   GenAI, `extract_text`) concatenate parts without a separator. Media-
 *   bearing results get the note as their own trailing text part.
 *
 * Callers: the live LLM projection (`agent/context/projector.ts`) and the
 * vis debugger's model view, which must mirror the live projection exactly.
 */
import type { ContentPart } from '@moonshot-ai/kosong';

export const TOOL_ERROR_STATUS = 'ERROR: Tool execution failed.';
export const TOOL_EMPTY_STATUS = 'Tool output is empty.';
export const TOOL_EMPTY_ERROR_STATUS = 'ERROR: Tool execution failed. Tool output is empty.';

export interface RenderableToolResult {
  readonly output: string | readonly ContentPart[];
  readonly note?: string | undefined;
  readonly isError?: boolean | undefined;
}

export function renderToolResultForModel(result: RenderableToolResult): ContentPart[] {
  const rendered = renderStatus(result);
  if (result.note === undefined || result.note.length === 0) {
    return rendered;
  }
  const only = rendered[0];
  if (rendered.length === 1 && only?.type === 'text') {
    return [textPart(`${only.text}\n${result.note}`)];
  }
  return [...rendered, textPart(result.note)];
}

function renderStatus(result: RenderableToolResult): ContentPart[] {
  const output = result.output;

  // String outputs — and their history form, a single text part — keep the
  // legacy joined shape: the status prefix shares one text part with the
  // output so provider serialization is unchanged.
  const single = typeof output === 'string' ? output : singleTextPart(output);
  if (single !== undefined) {
    if (result.isError === true) {
      if (single.length === 0) return [textPart(TOOL_EMPTY_ERROR_STATUS)];
      if (single.trimStart().startsWith('ERROR:')) return [textPart(single)];
      return [textPart(`${TOOL_ERROR_STATUS}\n${single}`)];
    }
    return isEmptyOutputText(single) ? [textPart(TOOL_EMPTY_STATUS)] : [textPart(single)];
  }

  const parts = output as readonly ContentPart[];
  // An array with no sendable content (empty, or only empty/whitespace-only
  // text blocks) gets the placeholder. Otherwise projection would drop the
  // blank text blocks, leave the tool message empty, and throw on every send.
  if (isEmptyEquivalentContentArray(parts)) {
    return [textPart(result.isError === true ? TOOL_EMPTY_ERROR_STATUS : TOOL_EMPTY_STATUS)];
  }
  if (result.isError === true) {
    return [textPart(TOOL_ERROR_STATUS), ...parts];
  }
  return [...parts];
}

function singleTextPart(output: readonly ContentPart[]): string | undefined {
  const first = output[0];
  return output.length === 1 && first?.type === 'text' ? first.text : undefined;
}

function textPart(text: string): ContentPart {
  return { type: 'text', text };
}

function isEmptyOutputText(output: string): boolean {
  return output.trim().length === 0 || output.trim() === TOOL_EMPTY_STATUS;
}

function isEmptyEquivalentContentArray(output: readonly ContentPart[]): boolean {
  return output.every((part) => part.type === 'text' && part.text.trim().length === 0);
}

File: packages/kimi-agent/src/context/tool_result_render.rs
/// Tool result rendering for model consumption.
///
/// Corresponds to `packages/agent-core/src/agent/context/tool-result-render.ts`.

use crate::context::types::ContentPart;

/// Status markers for tool results.
pub const TOOL_ERROR_STATUS: &str = "<system>ERROR: Tool execution failed.</system>";
pub const TOOL_EMPTY_STATUS: &str = "<system>Tool output is empty.</system>";
pub const TOOL_EMPTY_ERROR_STATUS: &str =
    "<system>ERROR: Tool execution failed. Tool output is empty.</system>";

/// Input for rendering a tool result for the model.
pub struct RenderableToolResult {
    pub output: ToolResultOutput,
    pub note: Option<String>,
    pub is_error: bool,
}

/// Tool result output: either a plain string or content parts.
pub enum ToolResultOutput {
    Text(String),
    Parts(Vec<ContentPart>),
}

/// Render a tool result for model consumption.
/// Applies error status prefix, empty-output placeholder, and trailing note.
pub fn render_tool_result_for_model(result: RenderableToolResult) -> Vec<ContentPart> {
    let rendered = render_status(&result);
    match &result.note {
        Some(note) if !note.is_empty() => {
            // Append note as a trailing text part.
            let mut out = rendered;
            if let Some(first) = out.first() {
                if out.len() == 1 && matches!(first, ContentPart::Text { .. }) {
                    // Join note into the single text part.
                    if let ContentPart::Text { text } = first {
                        return vec![ContentPart::Text {
                            text: format!("{text}\n{note}"),
                        }];
                    }
                }
            }
            out.push(ContentPart::Text {
                text: note.clone(),
            });
            out
        }
        _ => rendered,
    }
}

fn render_status(result: &RenderableToolResult) -> Vec<ContentPart> {
    let single = match &result.output {
        ToolResultOutput::Text(s) => Some(s.clone()),
        ToolResultOutput::Parts(parts) => single_text_part(parts),
    };

    if let Some(text) = single {
        if result.is_error {
            if text.is_empty() {
                return vec![ContentPart::Text {
                    text: TOOL_EMPTY_ERROR_STATUS.to_string(),
                }];
            }
            return vec![ContentPart::Text {
                text: format!("{TOOL_ERROR_STATUS}\n{text}"),
            }];
        }
        if is_empty_output_text(&text) {
            return vec![ContentPart::Text {
                text: TOOL_EMPTY_STATUS.to_string(),
            }];
        }
        return vec![ContentPart::Text { text }];
    }

    // Content parts array
    let parts = match &result.output {
        ToolResultOutput::Parts(p) => p,
        _ => unreachable!(),
    };

    if is_empty_equivalent_content_array(parts) {
        return vec![ContentPart::Text {
            text: if result.is_error {
                TOOL_EMPTY_ERROR_STATUS.to_string()
            } else {
                TOOL_EMPTY_STATUS.to_string()
            },
        }];
    }

    if result.is_error {
        let mut out = vec![ContentPart::Text {
            text: TOOL_ERROR_STATUS.to_string(),
        }];
        out.extend(parts.iter().cloned());
        return out;
    }

    parts.clone()
}

fn single_text_part(parts: &[ContentPart]) -> Option<String> {
    if parts.len() == 1 {
        if let ContentPart::Text { text } = &parts[0] {
            return Some(text.clone());
        }
    }
    None
}

fn is_empty_output_text(text: &str) -> bool {
    text.trim().is_empty() || text.trim() == "Tool output is empty."
}

fn is_empty_equivalent_content_array(parts: &[ContentPart]) -> bool {
    parts.iter().all(|part| {
        if let ContentPart::Text { text } = part {
            text.trim().is_empty()
        } else {
            false
        }
    })
}